src/hub.js

```diff
@@ -398,11 +410,108 @@ document.addEventListener("DOMContentLoaded", () => {
       console.error(res);
     });
 
+  const hubPhxPresence = new Presence(hubPhxChannel);
+  const presenceLogEntries = [];
+
+  const addToPresenceLog = entry => {
+    entry.key = new Date().getTime().toString();
+    presenceLogEntries.push(entry);
+    remountUI({ presenceLogEntries });
+
+    // Fade out and then remove
+    setTimeout(() => {
+      entry.expired = true;
+      remountUI({ presenceLogEntries });
+
+      setTimeout(() => {
+        presenceLogEntries.splice(presenceLogEntries.indexOf(entry), 1);
+        remountUI({ presenceLogEntries });
+      }, 5000);
+    }, entryManager.hasEntered() ? 10000 : 30000); // Fade out things faster once entered.
+  };
+  window.add = addToPresenceLog;
+
+  let isInitialSync = true;
+  let presences = {};
+
+  hubPhxPresence.onSync(() => {
+    if (isInitialSync) {
+      hubPhxPresence.onJoin((sessionId, current, info) => {
+        if (current) return;
+
+        const meta = info.metas[0];
+        // Wire up join/leave event handlers after initial sync.
+
+        if (meta.presence && meta.profile.displayName) {
+          addToPresenceLog({
+            type: "join",
+            presence: meta.presence,
+            name: meta.profile.displayName
+          });
+        }
+      });
+
+      hubPhxPresence.onLeave((sessionId, current, info) => {
+        if (current && current.metas.length > 0) return;
```

Comment:
i can clarify this -- i also need to confirm I understand what is going on here. when phoenix sends a `presence_diff` message, it includes `joins` and `leaves` -- `joins` are new presence k/v pairs and `leaves` are ones to discard. however, since this is for `presence_diff`, it also does this when the presence state for a key is mutated. (there's no `presence_update` event afaik.) so if I change a person's avatar name, it's sent over as a join and a leave in a single diff, and hence `onLeave` is called. so, the way to determine if a call to `onLeave` actually constitutes a real departure from the channel is to check to see if there is metadata associated with the current entry at all -- in which case it means that this was a mutation, not a departure. i will look at phoenix.js to make sure i'm not missing something though.

Reply:
ah yeah confirmed -- https://github.com/gfodor/phoenix-js/blob/upstream-phoenix-js/js/phoenix.js#L174

Reply:
upon review, I realized I could simplify this code -- see https://github.com/mozilla/hubs/commit/c98134ed2df170e7c37fd2131940284c24d6fbe0